File: apps/apple/SocialWire/Utilities/SavedLinkEmbedURL.swift
import Foundation

enum SavedLinkEmbedURL {
    static func isPocketReaderHostname(_ hostname: String) -> Bool {
        let h = hostname.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        return h == "getpocket.com"
            || h.hasSuffix(".getpocket.com")
            || h == "readitlaterlist.com"
            || h.hasSuffix(".pocket.com")
            || h == "pckt.it"
            || h == "pkt.cool"
    }

    static func isPoorIframeEmbedTarget(_ rawURL: String) -> Bool {
        guard let url = URL(string: rawURL.trimmingCharacters(in: .whitespacesAndNewlines)),
              let host = url.host
        else { return false }
        return isPocketReaderHostname(host)
    }

    static func resolveEmbedURL(for save: MergedLatrSave) -> String? {
        let linked = save.linkedWebUrl?.trimmingCharacters(in: .whitespacesAndNewlines)
        let primary: String? = {
            switch save {
            case .external(let row):
                return row.url.trimmingCharacters(in: .whitespacesAndNewlines)
            case .native(let row):
                return row.url?.trimmingCharacters(in: .whitespacesAndNewlines)
            }
        }()

        if let primary, !primary.isEmpty, isPoorIframeEmbedTarget(primary) {
            if let linked, !linked.isEmpty { return linked }
            return primary
        }
        if let primary, !primary.isEmpty { return primary }
        if let linked, !linked.isEmpty { return linked }
        return nil
    }

    static func previewURL(for save: MergedLatrSave) -> URL? {
        guard let raw = resolveEmbedURL(for: save) else { return nil }
        let normalized = PublicURLNormalizer.normalizeHttpURLToHTTPS(raw)
        return URL(string: normalized)
    }
}

File: apps/web/src/lib/savedLinkEmbedUrl.ts
import type { MergedLatrSave } from "@/lib/pdsClient";
import { sanitizeEmbedUrlForIframe } from "@/lib/publicResourceUrl";

/** Pocket reader / redirect hosts that loop or auth-fail inside sandboxed iframes. */
export function isPocketReaderHostname(hostname: string): boolean {
  const h = hostname.trim().toLowerCase();
  return (
    h === "getpocket.com" ||
    h.endsWith(".getpocket.com") ||
    h === "readitlaterlist.com" ||
    h.endsWith(".pocket.com")
  );
}

function isPocketReaderWrapperUrl(url: URL): boolean {
  if (!isPocketReaderHostname(url.hostname)) return false;
  const path = url.pathname.toLowerCase();
  return (
    path === "/read" ||
    path.startsWith("/read/") ||
    path === "/saved" ||
    path.startsWith("/saved/")
  );
}

export function isPoorIframeEmbedTarget(rawUrl: string): boolean {
  try {
    return isPocketReaderWrapperUrl(new URL(rawUrl.trim()));
  } catch {
    return false;
  }
}

/**
 * Canonical HTTPS URL to load in the saved-link reader iframe.
 * Prefers `linkedWebUrl` when the stored wrapper is a Pocket reader redirect.
 */
export function resolveSavedLinkEmbedUrl(row: MergedLatrSave): string | undefined {
  const linked = row.linkedWebUrl?.trim();
  const primary =
    row.kind === "external" ? row.url.trim() : row.url?.trim() ?? undefined;

  if (primary && isPoorIframeEmbedTarget(primary)) {
    return linked ?? primary;
  }
  return primary ?? linked;
}

export function stableSavedLinkIframeSrc(row: MergedLatrSave): string | undefined {
  const resolved = resolveSavedLinkEmbedUrl(row);
  if (!resolved) return undefined;
  return sanitizeEmbedUrlForIframe(resolved);
}
